Files/warranty_meraki.py

```diff
@@ -0,0 +1,343 @@
+                    if result['status'] == 'License Required':
+                        days_remaining = datetime.utcnow().strftime("%Y-%m-%d")
+                    else:
+                        days_remaining = datetime.utcnow().strftime("%Y-%m-%d")
+                else:
+                    days_remaining = self.meraki_date_parser(result['expirationDate'])  # TODO make sure this works
+
+                all_license_states[organization_id] = days_remaining
+
+        return all_license_states
+
+    def get_all_organizations(self, retry, timeout):
+        all_organization_ids = []
+
+        headers = {
+            'X-Cisco-Meraki-API-Key': self.api_key
+        }
+
+        if self.debug is True:
+            print '\t\t[+] Getting all organization IDs '
+
+        try:
+            resp = requests.get(self.url + self.organization_endpoint, headers=headers, timeout=timeout)
+            result = self.process_get_response(resp, retry)
+        except requests.RequestException as e:
+            self.error_msg(e)
+            return None
+
+        if result is None:
+            return None
+
+        # adds all the organization ids retrieved from the request to a combined list
+        for organization in result:
+            if 'id' in organization:
+                all_organization_ids.append(organization['id'])
+
+        return all_organization_ids
+
+    def get_all_device_serial_numbers(self, all_organization_ids, retry, timeout):
+        all_organizations = {}
+        devices_in_network = []
+
+        all_devices = {}
+
+        headers = {
+            'X-Cisco-Meraki-API-Key': self.api_key
+        }
+
+        if self.debug is True:
+            print '\t\t[+] Getting all device serial numbers '
+
+        for organization_id in all_organization_ids:
+            try:
+                resp = requests.get(self.url + self.organization_endpoint + '/' + organization_id + '/' +
+                                    self.devices_endpoint, headers=headers, timeout=timeout)
+                result = self.process_get_response(resp, retry)
+            except requests.RequestException as e:
+                self.error_msg(e)
+                continue
+
+            if result is None:
+                return None
+
+            # adds all the devices retrieved from the request to a dictionary with key serial number, value organization
+            for device in result:
+                if 'serial' in device:
+                    all_devices[device['serial']] = organization_id  # add device to all devices dictionary for later
+
+            # adds a list of all the devices in an organization to a dictionary
+            all_organizations[organization_id] = result
+
+        return all_organizations, all_devices
+
+    # TODO: add rate limit handling
+    def process_get_response(self, resp, retry):
+        msg = 'Status code: %s' % str(resp.status_code)
+
+        if str(resp.status_code) == '429' or str(resp.status_code) == '404':
```

Comment:
Are these the status codes that Meraki API returns that mean "try again"?

Reply:
429 means too many requests, 404 means that the user may not have permission to make api calls to an organization, but may be a result of a malformed request. added an api throttler to prevent 429 but keeping this section in case that fails for some unknown reason